koku/subs/subs_data_extractor.py

```diff
@@ -0,0 +1,170 @@
+    def __init__(self, schema, provider_type, provider_uuid, tracing_id, context=None):
+        super().__init__(schema)
+        self.provider_type = provider_type.removesuffix("-local")
+        self.provider_uuid = provider_uuid
+        self.tracing_id = tracing_id
+        self.table = TABLE_MAP.get(self.provider_type)
+        self.s3_client = get_subs_s3_client()
+        self.context = context
+
+    @cached_property
+    def subs_s3_path(self):
+        """The S3 path to be used for a SUBS report upload."""
+        return f"{self.schema}/{self.provider_type}/source={self.provider_uuid}/date={self.date_helper.today.date()}"
+
+    def determine_latest_processed_time_for_provider(self, year, month):
+        """Determine the latest processed timestamp for a provider for a given month and year."""
+        with schema_context(self.schema):
+            last_time = SubsLastProcessed.objects.filter(
+                source_uuid=self.provider_uuid, year=year, month=month
+            ).first()
+        return last_time.latest_processed_time if last_time else None
+
+    def determine_end_time(self, year, month):
+        sql = (
+            f" SELECT MAX(lineitem_usagestartdate) FROM aws_line_items"
+            f" WHERE source='{self.provider_uuid}' AND year='{year}' AND month='{month}'"
+        )
+        latest = self._execute_trino_raw_sql_query(sql, log_ref="insert_subs_last_processed_time")
+        return latest[0][0]
+
+    def determine_line_item_count(self, where_clause):
+        """Determine the number of records in the table that have not been processed and match the criteria"""
+        table_count_sql = f"SELECT count(*) FROM {self.schema}.{self.table} {where_clause}"
+        count = self._execute_trino_raw_sql_query(table_count_sql, log_ref="determine_subs_processing_count")
+        return count[0][0]
+
+    def determine_where_clause(self, latest_processed_time, end_time, year, month):
+        """Determine the where clause to use when processing subs data"""
+        return (
+            f"WHERE source='{self.provider_uuid}' AND year='{year}' AND month='{month}' AND"
+            " lineitem_productcode = 'AmazonEC2' AND lineitem_lineitemtype IN ('Usage', 'SavingsPlanCoveredUsage') AND"
+            " product_vcpu IS NOT NULL AND strpos(resourcetags, 'com_redhat_rhel') > 0 AND"
+            f" lineitem_usagestartdate > TIMESTAMP '{latest_processed_time}' AND"
+            f" lineitem_usagestartdate <= TIMESTAMP '{end_time}'"
+        )
+
+    def update_latest_processed_time(self, year, month, end_time):
+        """Update the latest processing time for a provider"""
+        with schema_context(self.schema):
+            subs_obj = SubsLastProcessed.objects.filter(
+                source_uuid=self.provider_uuid, year=year, month=month
+            ).first() or SubsLastProcessed.objects.create(
+                source_uuid=TenantAPIProvider.objects.get(uuid=self.provider_uuid), year=year, month=month
+            )
```

Comment:
I was thinking about the times we had run into timing issues with get_or_create in the past, but since I made the `source_uuid`, `year` and `month` unique together, that should be fine. Updated to use `get_or_create` here.

As for the `TenantAPIProvider` lookup, I had forgotten that inserting with Foreign keys using django models was different and `source_uuid` for the insert required a reference to the object but `source_uuid_id` could be the matching UUID. I have updated that accordingly.